From ee074d298e8151cb494c3f841236c8db2812e968 Mon Sep 17 00:00:00 2001
From: Simeon Marijon <simeon.marijon@foss.st.com>
Date: Fri, 1 Dec 2023 13:42:57 +0100
Subject: [PATCH 436/841] feat(stm32mp2): initialize TAMP Backup Registers for
 BL2 and BL31

After the init call the driver stm32_tamp_nvram will be
registered and so the nvmem cells description in the devicetree
will be accessible.

In BL31, we redo the initialisation, as the RIF configuration
should have been updated by OPTEE. We have a verified access
to TAMP Backup registers at the very end of the initialisation of
BL31 (at bl31_plat_runtime_setup)

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 8c1e62c4ff7233d5bf54fe73112acedf272d3dd3 feat(stm32mp2): initialize TAMP Backup Registers for BL2 and BL31]
Signed-off-by: Simeon Marijon <simeon.marijon@foss.st.com>
Change-Id: I1f8a2be699b4d3cb296d3f04264241c980ebaf47
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/343951
Tested-by: Simeon MARIJON <simeon.marijon@st.com>
Reviewed-by: Simeon MARIJON <simeon.marijon@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp2/bl31_plat_setup.c | 16 ++++++++++++++++
 plat/st/stm32mp2/stm32mp2_def.h    |  2 +-
 2 files changed, 17 insertions(+), 1 deletion(-)

diff --git a/plat/st/stm32mp2/bl31_plat_setup.c b/plat/st/stm32mp2/bl31_plat_setup.c
index 425409706..75dc1f32a 100644
--- a/plat/st/stm32mp2/bl31_plat_setup.c
+++ b/plat/st/stm32mp2/bl31_plat_setup.c
@@ -61,6 +61,8 @@ void bl31_early_platform_setup2(u_register_t arg0, u_register_t arg1,
 		panic();
 	}
 
+	stm32_tamp_nvram_init();
+
 	(void)stm32mp_uart_console_setup();
 
 	assert(arg0 != 0UL);
@@ -133,6 +135,20 @@ void bl31_platform_setup(void)
 {
 }
 
+void bl31_plat_runtime_setup(void)
+{
+	/* We reinit the nvram driver Optee should have update the register access configuration
+	 * based on the device tree
+	 */
+	int ret = stm32_tamp_nvram_update_rights();
+
+	if (ret != 0) {
+		ERROR("Failed to update TAMP backup registers rights\n");
+		panic();
+	}
+	console_switch_state(CONSOLE_FLAG_RUNTIME);
+}
+
 entry_point_info_t *bl31_plat_get_next_image_ep_info(unsigned int type)
 {
 	if (type == NON_SECURE)
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 9a28d16a6..acaa37c80 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -140,7 +140,7 @@ enum ddr_type {
 						 (STM32MP_PARAM_LOAD_SIZE +	\
 						  STM32MP_HEADER_SIZE))
 
-#define STM32MP_BL31_SIZE		U(0x0001D000)	/* 116 KB for BL31 */
+#define STM32MP_BL31_SIZE		U(0x0001F000)	/* 124 KB for BL31 */
 
 #if TRUSTED_BOARD_BOOT
 #define STM32MP_BL2_RO_SIZE		U(0x00021000)	/* 132 KB */
-- 
2.47.2

